pkg/service/user/user.go

```diff
@@ -15,20 +15,29 @@ import (
 	"github.com/tensorchord/envd-server/pkg/query"
 )
 
-type Service struct {
+type Service interface {
+	Register(loginName, pwd string,
+		PublicKey []byte) (string, error)
+	GetPubKey(loginName string) ([]byte, error)
+	Login(loginName, pwd string, auth bool) (bool, string, error)
+
+	ValidateJWT(token string) (string, error)
+}
+
+type generalService struct {
 	querier   query.Querier
 	jwtIssuer *JWTIssuer
 }
 
 func NewService(querier query.Querier,
-	secret string, expirationTimeDefault time.Duration) *Service {
-	return &Service{
+	secret string, expirationTimeDefault time.Duration) Service {
+	return &generalService{
```

Comment:
So it's a pointer but then convert into the value?

Reply:
It's the pointer to the struct `generalService`, which implements the interface `Service`

Reply:
I see. It has to be a pointer so the compiler will know whether it implemented the interface or not. Golang syntax is weird.